File: bytes/bench/bench.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <time.h>

static double now_ms(void) {
	struct timespec ts;
	clock_gettime(CLOCK_MONOTONIC, &ts);
	return ts.tv_sec * 1000.0 + ts.tv_nsec / 1e6;
}

static uint32_t fill_buf(uint8_t *b, size_t n) {
	for (size_t i = 0; i < n; i++)
		b[i] = (uint8_t)((i * 31 + 7) & 255);
	b[n - 4] = 13;
	b[n - 3] = 10;
	b[n - 2] = 13;
	b[n - 1] = 10;
	return (uint32_t)b[12345] + (uint32_t)b[n - 1];
}

static int build(size_t n) {
	uint8_t *out = NULL;
	size_t len = 0, cap = 0;
	double t0 = now_ms();
	for (size_t i = 0; i < n; i++) {
		if (len == cap) {
			size_t next = cap ? cap * 2 : 1;
			uint8_t *grown = realloc(out, next);
			if (!grown) { free(out); return 1; }
			out = grown;
			cap = next;
		}
		out[len++] = (uint8_t)(i & 255);
	}
	uint32_t cs = (uint32_t)n + out[len - 1];
	printf("build_%zu\t%.6f\t%u\n", n, now_ms() - t0, cs);
	free(out);
	return 0;
}

int main(int argc, char **argv) {
	int L = 28;
	if (argc > 1)
		L = atoi(argv[1]);
	size_t n = (size_t)1 << L;

	double t0, t1;
	uint32_t cs;

	t0 = now_ms();
	uint8_t *b = malloc(n);
	if (!b) return 1;
	cs = fill_buf(b, n);
	t1 = now_ms();
	printf("fill\t%.6f\t%u\n", t1 - t0, cs);

	t0 = now_ms();
	uint32_t sum = 0;
	for (size_t i = 0; i < n; i++)
		sum += b[i];
	t1 = now_ms();
	printf("sum\t%.6f\t%u\n", t1 - t0, sum);

	static const uint8_t needle[] = {13, 10, 13, 10};
	t0 = now_ms();
	void *p = memmem(b, n, needle, 4);
	t1 = now_ms();
	size_t find_idx = p ? (size_t)((uint8_t *)p - b) : (size_t)-1;
	printf("find\t%.6f\t%u\n", t1 - t0, (uint32_t)find_idx);

	size_t slice_start = n / 4 + 1;
	size_t slice_len = n / 2;
	t0 = now_ms();
	uint8_t *s = malloc(slice_len);
	memcpy(s, b + slice_start, slice_len);
	cs = (uint32_t)s[0] + (uint32_t)s[slice_len - 1];
	t1 = now_ms();
	printf("slice\t%.6f\t%u\n", t1 - t0, cs);
	free(s);

	t0 = now_ms();
	uint8_t *out = NULL;
	size_t out_len = 0;
	size_t out_cap = 0;
	size_t chunks = n / 65536;
	for (size_t k = 0; k < chunks; k++) {
		if (out_len + 65536 > out_cap) {
			size_t new_cap = out_cap ? out_cap * 2 : 65536;
			while (new_cap < out_len + 65536)
				new_cap *= 2;
			uint8_t *tmp = realloc(out, new_cap);
			if (!tmp) return 1;
			out = tmp;
			out_cap = new_cap;
		}
		uint8_t byte = (uint8_t)(k & 255);
		memset(out + out_len, byte, 65536);
		out_len += 65536;
	}
	t1 = now_ms();
	if (out_len != n)
		return 1;
	cs = (uint32_t)out[0] + (uint32_t)out[out_len - 1];
	printf("concat\t%.6f\t%u\n", t1 - t0, cs);
	free(out);

	uint32_t x = 1, acc = 0;
	unsigned shift = 32 - (unsigned)L;
	t0 = now_ms();
	for (uint32_t i = 0; i < (1u << 24); i++) {
		x = x * 1664525u + 1013904223u;
		uint32_t idx = x >> shift;
		acc += b[idx];
	}
	t1 = now_ms();
	printf("random\t%.6f\t%u\n", t1 - t0, acc);

	uint8_t *c = malloc(n);
	memcpy(c, b, n);
	t0 = now_ms();
	int eq = memcmp(b, c, n) == 0;
	t1 = now_ms();
	printf("equal\t%.6f\t%u\n", t1 - t0, eq ? 1u : 0u);

	if (build(1000000) || build(4000000))
		return 1;

	free(c);
	free(b);
	return 0;
}

File: bytes/bench/bench.rs
use std::hint::black_box;
use std::time::Instant;

fn fill_buf(b: &mut [u8]) -> u32 {
    for (i, byte) in b.iter_mut().enumerate() {
        *byte = ((i as u64 * 31 + 7) & 255) as u8;
    }
    let n = b.len();
    b[n - 4] = 13;
    b[n - 3] = 10;
    b[n - 2] = 13;
    b[n - 1] = 10;
    b[12345] as u32 + b[n - 1] as u32
}

fn main() {
    let l: u32 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(28);
    let n = 1usize << l;

    let t0 = Instant::now();
    let mut b = vec![0u8; n];
    let cs = fill_buf(&mut b);
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    println!("fill\t{ms:.6}\t{cs}");

    let t0 = Instant::now();
    let mut sum: u32 = 0;
    for &byte in &b {
        sum = sum.wrapping_add(byte as u32);
    }
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    println!("sum\t{ms:.6}\t{sum}");

    let needle = [13u8, 10, 13, 10];
    let t0 = Instant::now();
    let find_idx = b
        .windows(4)
        .position(|w| w == needle)
        .unwrap_or(usize::MAX);
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    println!("find\t{ms:.6}\t{}", find_idx as u32);

    let slice_start = n / 4 + 1;
    let slice_len = n / 2;
    let t0 = Instant::now();
    let s: Vec<u8> = b[slice_start..slice_start + slice_len].to_vec();
    let cs = s[0] as u32 + s[s.len() - 1] as u32;
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    println!("slice\t{ms:.6}\t{cs}");

    let t0 = Instant::now();
    let mut out: Vec<u8> = Vec::new();
    let chunks = n / 65536;
    for k in 0..chunks {
        let byte = (k & 255) as u8;
        out.extend(std::iter::repeat(byte).take(65536));
    }
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    if out.len() != n {
        std::process::exit(1);
    }
    let cs = out[0] as u32 + out[out.len() - 1] as u32;
    println!("concat\t{ms:.6}\t{cs}");

    let mut x: u32 = 1;
    let mut acc: u32 = 0;
    let shift = 32 - l;
    let t0 = Instant::now();
    for _ in 0..(1u32 << 24) {
        x = x.wrapping_mul(1664525).wrapping_add(1013904223);
        let idx = (x >> shift) as usize;
        acc = acc.wrapping_add(b[idx] as u32);
    }
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    black_box(acc);
    println!("random\t{ms:.6}\t{acc}");

    let c = b.clone();
    let t0 = Instant::now();
    let eq = b == c;
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    println!("equal\t{ms:.6}\t{}", if eq { 1u32 } else { 0 });

    black_box(b);
}
